CHANGELOG.md

```diff
@@ -1,8 +1,487 @@
+establish hundreds of connections, significantly impacting the performance of
+go-ipfs and [crashing some
+routers](https://github.com/ipfs/go-ipfs/issues/3320).
+
+We've introduced a new persistent provider queue for files added via `ipfs add`
+and `ipfs pin add`. When new directory trees are added to go-ipfs, go-ipfs will
+add the root/final CID to this queue. Then, in the background, go-ipfs will walk
+the queue, sequentially sending out provider records for each CID.
+
+This ensures that root CIDs are sent out as soon as possible and are sent even
+when files are added when the go-ipfs daemon isn't running.
+
+By example, let's add a directory tree to go-ipfs:
+
+```bash
+> # We're going to do this in "online" mode first so let's start the daemon.
+> ipfs daemon &
+...
+Daemon is ready
+> # Now, we're going to create a directory to add.
+> mkdir foo
+> for i in {0..1000}; do echo do echo $i > foo/$i; done
+> # finally, we're going to add it.
+> ipfs add -r foo
+added QmUQcSjQx2bg4cSe2rUZyQi6F8QtJFJb74fWL7D784UWf9 foo/0
+...
+added QmQac2chFyJ24yfG2Dfuqg1P5gipLcgUDuiuYkQ5ExwGap foo/990
+added QmQWwz9haeQ5T2QmQeXzqspKdowzYELShBCLzLJjVa2DuV foo/991
+added QmQ5D4MtHUN4LTS4n7mgyHyaUukieMMyCfvnzXQAAbgTJm foo/992
+added QmZq4n4KRNq3k1ovzxJ4qdQXZSrarfJjnoLYPR3ztHd7EY foo/993
+added QmdtrsuVf8Nf1s1MaSjLAd54iNqrn1KN9VoFNgKGnLgjbt foo/994
+added QmbstvU9mnW2hsE94WFmw5WbrXdLTu2Sf9kWWSozrSDscL foo/995
+added QmXFd7f35gAnmisjfFmfYKkjA3F3TSpvUYB9SXr6tLsdg8 foo/996
+added QmV5BxS1YQ9V227Np2Cq124cRrFDAyBXNMqHHa6kpJ9cr6 foo/997
+added QmcXsccUtwKeQ1SuYC3YgyFUeYmAR9CXwGGnT3LPeCg5Tx foo/998
+added Qmc4mcQcpaNzyDQxQj5SyxwFg9ZYz5XBEeEZAuH4cQirj9 foo/999
+added QmXpXzUhcS9edmFBuVafV5wFXKjfXkCQcjAUZsTs7qFf3G foo
+```
+
+In 0.4.19, we would have sent out provider records for files `foo/{0..1000}`
+_before_ sending out a provider record for `foo`. If you were ask a friend to
+download /ipfs/QmUQcSjQx2bg4cSe2rUZyQi6F8QtJFJb74fWL7D784UWf9, they would
+(baring other issues) be able to find it pretty quickly as this is the first CID
+you'll have announced to the network. However, if you ask your friend to
+download /ipfs/QmXpXzUhcS9edmFBuVafV5wFXKjfXkCQcjAUZsTs7qFf3G/0, they'll have to
+wait for you to finish telling the network about every file in `foo` first.
+
+In 0.4.20, we _immediately_ tell the network about
+`QmXpXzUhcS9edmFBuVafV5wFXKjfXkCQcjAUZsTs7qFf3G` (the `foo` directory) as soon
+as we finish adding the directory to go-ipfs _without_ waiting to finish
+announcing `foo/{0..1000}`. This is especially important in this release
+because we've drastically reduced the number of provide workers.
+
+The second benefit is that this queue is persistent. That means go-ipfs won't
+forget to send out this record, even if it was offline when the content was
+initially added. NOTE: go-ipfs _does_ continuously _re_-send provider records in
+the background twice a day, it just might be a while before it gets around to
+sending one out any specific one.
```

Comment:
typo (very minor): I think this should be something like "...before it gets around to sending any specific one"

Reply:
(could you file a PR to fix this)